cEP-0016.md

```diff
@@ -0,0 +1,234 @@
+
+```python
+    if len(shortlog) > shortlog_length:
+        github_merge_commit_regex = (^Merge [0-9a-f]+ into [0-9a-f]+$|....)
+
+        match = re.fullmatch(github_merge_commit_regex, shortlog)
+        if match:
+            # ignore this commit from any further inspection
+            # inspect the unmerged parent of this commit instead
+```
+
+A prototype for identifying merge commits:
+
+```python
+    def is_merge_commit(self, shortlog, body):
+        commit_hash = run_shell_command('git rev-parse HEAD')
+
+        command = 'git show --no-patch --format="%P" '
+        command += str(commit_hash)
+
+        commit_parents = run_shell_command(command)
+
+        if commit_parents > 1:
+            return 1
+        else:
+            return 0
+```
+
+A prototype for identifying commits that disable ci builds:
+
+```python
+    def check_skip_cibuild(self, shortlog, body):
+        ci_regex = re.compile(r'ci skip')
+
+        if ci_regex.match(shortlog):
+            # Here we would check all of the files modified
+            # to ensure that 'ci skip' is permitted
+```
+
+A prototype for identifying revert commits:
+
+```python
+    def is_revert_commit(self, shortlog, body):
+        revert_desc_regex = re.compile(r'^Revert\s"[\S+\s*]*"')
+        headline_regex = re.compile(
+            r'^This reverts commit [0-9a-f]{40}')
+        sha_regex = re.compile(r'\b[0-9a-f]{40}\b')
+
+        if not allow_revert_commits and revert_desc_regex.match(shortlog):
+            reverted_shortlog = shortlog.lstrip('Revert ')
+
+            if len(body) != 0:
+                body = body.strip('\n')
+                blankline = body.find('\n\n')
+                headline = body[:blankline] if blankline != -1 else body
+
+                if headline_regex.match(headline):
+                    commit_sha = re.search(sha_regex, headline)
+                    commit_sha = commit_sha.group(0)
+                    command = 'git log -n 1 ' + str(commit_sha)
+                    rstdout, rstderr = run_shell_command(command)
+
+                    if rstderr:
+                        self.err('git:', repr(stderr))
+                        yield Result(self, 'Invalid revert commit.')
+
+                    rstdout = rstdout.rstrip('\n')
+                    rpos = rstdout.find('\n')
+                    old_shortlog = rstdout[:rpos] if rpos != -1 else rstdout
+
+                    if old_shortlog == reverted_shortlog:
+                        return 1
+
+                else:
+                    return 0
+```
+
+## Handling the special commits once identified
+
+The `GitCommitBear` checks the format of all the commits made to coala. With
+above changes, the special commits will be identified and treated separately
+in the following manner:
+
+Git merge commits
+
+```python
+    if not allow_merge_commit and is_merge_commit:
+        yield Result(self, 'Merge commits are not allowed. Please change your commit.')
+```
```

Comment:
how/where is defined `allow_merge_commit`? I guess it's from the `.coafile` but I don't *know* ☺